components/validator/challenger.go

```diff
@@ -65,191 +93,413 @@ func NewChallenger(ctx context.Context, cfg Config, l log.Logger) (*Challenger,
+
+	return nil
 }
 
-type OutputRange struct {
-	OutputIndex *big.Int
-	StartBlock  uint64
-	EndBlock    uint64
+func (c *Challenger) Stop() {
+	c.log.Info("stop challenger")
+	close(c.done)
+	c.wg.Wait()
 }
 
-func (c *Challenger) GetInvalidOutputRange() (*OutputRange, error) {
-	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(c.callOpts)
-	if err != nil {
-		return nil, err
-	}
-	latestOutputIndex := new(big.Int).Sub(nextOutputIndex, common.Big1)
-	if latestOutputIndex.Cmp(common.Big1) == -1 {
-		c.log.Info("challenge for genesis output is not allowed", "latestOutputIndex", latestOutputIndex)
-		return nil, nil
+// scanPrevOutputs scans all the previous outputs within the finalization window.
+// If there are invalid outputs, create challenge.
+// If there are challenges in progress, keep handling them.
+func (c *Challenger) scanPrevOutputs() {
+	defer c.wg.Done()
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), c.cfg.NetworkTimeout)
+	defer cancelCtx()
+	status, err := c.cfg.RollupClient.SyncStatus(ctx)
+
+	toBlock := new(big.Int).SetUint64(status.CurrentL1.Number)
+	// fromBlock = max(1, startBlockOfFinalizationWindow)
+	fromBlock := math.BigMax(new(big.Int).Sub(toBlock, new(big.Int).Div(c.finalizationPeriodSeconds, big.NewInt(12))), big.NewInt(1))
+
+	outputSubmittedEvent := c.l2ooABI.Events["OutputSubmitted"]
+	challengeCreatedEvent := c.colosseumABI.Events["ChallengeCreated"]
+
+	query := goEth.FilterQuery{
+		FromBlock: fromBlock,
+		ToBlock:   toBlock,
+		Addresses: []common.Address{c.cfg.L2OutputOracleAddr, c.cfg.ColosseumAddr},
+		Topics: [][]common.Hash{
+			{outputSubmittedEvent.ID, challengeCreatedEvent.ID},
+		},
 	}
 
-	if c.checkpoint == nil {
-		submissionIntervalSeconds := new(big.Int).Mul(c.submissionInterval, c.l2BlockTime)
-		waitingOutputNum := new(big.Int).Div(c.finalizationPeriodSeconds, submissionIntervalSeconds)
-		if latestOutputIndex.Cmp(waitingOutputNum) != 1 {
-			c.checkpoint = common.Big1
-		} else if waitingOutputNum.Cmp(common.Big0) == 0 {
-			c.checkpoint = latestOutputIndex
-			return nil, nil
-		} else {
-			c.checkpoint = new(big.Int).Sub(nextOutputIndex, waitingOutputNum)
-		}
+	logs, err := c.l1Client.FilterLogs(context.Background(), query)
+	if err != nil {
+		c.log.Error("failed to get event logs related to outputs", "err", err)
+		return
 	}
 
-	for i := c.checkpoint; i.Cmp(latestOutputIndex) != 1; i.Add(i, common.Big1) {
-		output, err := c.l2ooContract.GetL2Output(c.callOpts, i)
-		if err != nil {
-			return nil, err
-		}
+	for _, vLog := range logs {
+		// for OutputSubmitted event
+		if vLog.Address == c.cfg.L2OutputOracleAddr {
+			outputIndex := new(big.Int).SetBytes(vLog.Topics[2][:])
 
-		knownOutput, err := c.OutputAtBlockSafe(output.L2BlockNumber.Uint64(), false)
-		if err != nil {
-			return nil, err
+			outputRange, err := c.ValidateOutput(outputIndex)
+			if err != nil {
+				c.log.Warn("unable to validate output", "err", err, "outputIndex", outputIndex)
```

Comment:
This error happens only due to unresponsive L2 Geth or L2 nodes. So it can be helpful to record the output or relevant information when such errors occur. Then I think we should retry to call `ValidateOutput` once again.